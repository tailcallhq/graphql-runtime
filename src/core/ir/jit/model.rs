use std::fmt::{Debug, Formatter};

use crate::core::ir::model::IR;

#[allow(unused)]
#[derive(Debug, Clone)]
pub struct Arg {
    pub id: ArgId,
    pub name: String,
    pub type_of: crate::core::blueprint::Type,
    pub value: Option<async_graphql_value::Value>,
    pub default_value: Option<async_graphql_value::ConstValue>,
}

#[derive(Clone)]
pub struct ArgId(usize);

impl Debug for ArgId {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl ArgId {
    pub fn new(id: usize) -> Self {
        ArgId(id)
    }
}

#[derive(Clone, PartialEq, Eq, Hash)]
pub struct FieldId(usize);

impl Debug for FieldId {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl FieldId {
    pub fn new(id: usize) -> Self {
        FieldId(id)
    }
}

#[derive(Clone)]
pub struct Field<A: Clone> {
    pub id: FieldId,
    pub name: String,
    pub ir: Option<IR>,
    pub type_of: crate::core::blueprint::Type,
    pub args: Vec<Arg>,
    pub refs: Option<A>,
}

const EMPTY_VEC: &Vec<Field<Children>> = &Vec::new();
impl Field<Children> {
    #[allow(unused)]
    pub fn children(&self) -> &Vec<Field<Children>> {
        match &self.refs {
            Some(Children(children)) => children,
            _ => EMPTY_VEC,
        }
    }
}

impl Field<Parent> {
    fn parent(&self) -> Option<&FieldId> {
        self.refs.as_ref().map(|Parent(id)| id)
    }

    fn into_children(self, fields: &[Field<Parent>]) -> Field<Children> {
        let mut children = Vec::new();
        for field in fields.iter() {
            if let Some(id) = field.parent() {
                if *id == self.id {
                    children.push(field.to_owned().into_children(fields));
                }
            }
        }

        let refs = if children.is_empty() {
            None
        } else {
            Some(Children(children))
        };

        Field {
            id: self.id,
            name: self.name,
            ir: self.ir,
            type_of: self.type_of,
            args: self.args,
            refs,
        }
    }
}

impl<A: Debug + Clone> Debug for Field<A> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut debug_struct = f.debug_struct("Field");
        debug_struct.field("id", &self.id);
        debug_struct.field("name", &self.name);
        if self.ir.is_some() {
            debug_struct.field("ir", &"Some(..)");
        }
        debug_struct.field("type_of", &self.type_of);
        if !self.args.is_empty() {
            debug_struct.field("args", &self.args);
        }
        if self.refs.is_some() {
            debug_struct.field("refs", &self.refs);
        }
        debug_struct.finish()
    }
}

#[derive(Clone)]
pub struct Parent(FieldId);
#[allow(unused)]
impl Parent {
    pub fn new(id: FieldId) -> Self {
        Parent(id)
    }
}
impl Debug for Parent {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "Parent({:?})", self.0)
    }
}

#[derive(Clone, Debug)]
pub struct Children(Vec<Field<Children>>);

#[derive(Clone, Debug)]
pub struct ExecutionPlan {
    parent: Vec<Field<Parent>>,
    children: Vec<Field<Children>>,
}

impl ExecutionPlan {
    pub fn new(fields: Vec<Field<Parent>>) -> Self {
        let field_children = fields
            .clone()
            .into_iter()
            .filter(|f| f.refs.is_none())
            .map(|f| f.into_children(&fields))
            .collect::<Vec<_>>();

        Self { parent: fields, children: field_children }
    }

    #[allow(unused)]
    pub fn as_children(&self) -> &[Field<Children>] {
        &self.children
    }

    #[allow(unused)]
    pub fn as_parent(&self) -> &[Field<Parent>] {
        &self.parent
    }

    #[allow(unused)]
    pub fn find_field(&self, id: FieldId) -> Option<&Field<Parent>> {
        self.parent.iter().find(|field| field.id == id)
    }
}
