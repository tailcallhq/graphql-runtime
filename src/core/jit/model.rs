use std::collections::HashMap;
use std::fmt::{Debug, Formatter};

use async_graphql::parser::types::OperationType;
use async_graphql::Pos;
use serde::Deserialize;

use crate::core::ir::model::IR;

#[derive(Debug, Deserialize, Clone)]
pub struct Variables<Value>(HashMap<String, Value>);

impl<Value> Default for Variables<Value> {
    fn default() -> Self {
        Self::new()
    }
}

impl<Value> Variables<Value> {
    pub fn new() -> Self {
        Self(HashMap::new())
    }
    pub fn get(&self, key: &str) -> Option<&Value> {
        self.0.get(key)
    }
    pub fn insert(&mut self, key: String, value: Value) {
        self.0.insert(key, value);
    }
}

impl<V> FromIterator<(String, V)> for Variables<V> {
    fn from_iter<T: IntoIterator<Item = (String, V)>>(iter: T) -> Self {
        Self(iter.into_iter().collect())
    }
}

#[derive(Debug, Clone)]
pub struct Arg<Input> {
    pub id: ArgId,
    pub name: String,
    pub type_of: crate::core::blueprint::Type,
    pub value: Option<Input>,
    pub default_value: Option<Input>,
}

impl<Input> Arg<Input> {
    pub fn try_map<Output, Error>(
        self,
        map: impl Fn(Input) -> Result<Output, Error>,
    ) -> Result<Arg<Output>, Error> {
        Ok(Arg {
            id: self.id,
            name: self.name,
            type_of: self.type_of,
            value: self.value.map(&map).transpose()?,
            default_value: self.default_value.map(&map).transpose()?,
        })
    }
}

#[derive(Clone)]
pub struct ArgId(usize);

impl Debug for ArgId {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl ArgId {
    pub fn new(id: usize) -> Self {
        ArgId(id)
    }
}

#[derive(Clone, PartialEq, Eq, Hash)]
pub struct FieldId(usize);

impl Debug for FieldId {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl FieldId {
    pub fn new(id: usize) -> Self {
        FieldId(id)
    }
    pub fn as_usize(&self) -> usize {
        self.0
    }
}

#[derive(Clone)]
pub struct Field<Extensions, Input> {
    pub id: FieldId,
    pub name: String,
    pub ir: Option<IR>,
    pub type_of: crate::core::blueprint::Type,
    pub skip: Option<Variable>,
    pub include: Option<Variable>,
    pub args: Vec<Arg<Input>>,
    pub extensions: Option<Extensions>,
    pub pos: Pos,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Variable(String);

impl Variable {
    pub fn new(name: String) -> Self {
        Variable(name)
    }
    pub fn as_str(&self) -> &str {
        &self.0
    }
    pub fn into_string(self) -> String {
        self.0
    }
}

impl<Extensions, Input> Field<Extensions, Input> {
    pub fn try_map<Output, Error>(
        self,
        map: impl Fn(Input) -> Result<Output, Error>,
    ) -> Result<Field<Extensions, Output>, Error> {
        Ok(Field {
            id: self.id,
            name: self.name,
            ir: self.ir,
            type_of: self.type_of,
            extensions: self.extensions,
            skip: self.skip,
            include: self.include,
            pos: self.pos,
            args: self
                .args
                .into_iter()
                .map(|arg| arg.try_map(&map))
                .collect::<Result<_, _>>()?,
        })
    }
}

impl<Input> Field<Nested<Input>, Input> {
    pub fn nested(&self) -> Option<&Vec<Field<Nested<Input>, Input>>> {
        self.extensions.as_ref().map(|Nested(nested)| nested)
    }

    pub fn nested_iter(&self) -> impl Iterator<Item = &Field<Nested<Input>, Input>> {
        self.nested()
            .map(|nested| nested.iter())
            .into_iter()
            .flatten()
    }
}

impl<Input> Field<Flat, Input> {
    fn parent(&self) -> Option<&FieldId> {
        self.extensions.as_ref().map(|Flat(id)| id)
    }

    fn into_nested(self, fields: &[Field<Flat, Input>]) -> Field<Nested<Input>, Input>
    where
        Input: Clone,
    {
        let mut children = Vec::new();
        for field in fields.iter() {
            if let Some(id) = field.parent() {
                if *id == self.id {
                    children.push(field.to_owned().into_nested(fields));
                }
            }
        }

        let extensions = if children.is_empty() {
            None
        } else {
            Some(Nested(children))
        };

        Field {
            id: self.id,
            name: self.name,
            ir: self.ir,
            type_of: self.type_of,
            skip: self.skip,
            include: self.include,
            args: self.args,
            pos: self.pos,
            extensions,
        }
    }
}

impl<Extensions: Debug, Input: Debug> Debug for Field<Extensions, Input> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut debug_struct = f.debug_struct("Field");
        debug_struct.field("id", &self.id);
        debug_struct.field("name", &self.name);
        if self.ir.is_some() {
            debug_struct.field("ir", &"Some(..)");
        }
        debug_struct.field("type_of", &self.type_of);
        if !self.args.is_empty() {
            debug_struct.field("args", &self.args);
        }
        if self.extensions.is_some() {
            debug_struct.field("extensions", &self.extensions);
        }
        debug_struct.finish()
    }
}

/// Stores field relationships in a flat structure where each field links to its
/// parent.
#[derive(Clone)]
pub struct Flat(FieldId);

impl Flat {
    pub fn new(id: FieldId) -> Self {
        Flat(id)
    }
    pub fn as_id(&self) -> &FieldId {
        &self.0
    }
}
impl Debug for Flat {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "Flat({:?})", self.0)
    }
}

/// Store field relationships in a nested structure like a tree where each field
/// links to its children.
#[derive(Clone, Debug)]
pub struct Nested<Input>(Vec<Field<Nested<Input>, Input>>);

#[derive(Clone, Debug)]
pub struct OperationPlan<Input> {
    flat: Vec<Field<Flat, Input>>,
    operation_type: OperationType,
    nested: Vec<Field<Nested<Input>, Input>>,
}

impl<Input> OperationPlan<Input> {
    pub fn new(fields: Vec<Field<Flat, Input>>, operation_type: OperationType) -> Self
    where
        Input: Clone,
    {
        let nested = fields
            .clone()
            .into_iter()
            .filter(|f| f.extensions.is_none())
            .map(|f| f.into_nested(&fields))
            .collect::<Vec<_>>();

        Self { flat: fields, nested, operation_type }
    }

    pub fn operation_type(&self) -> OperationType {
        self.operation_type
    }

    pub fn is_query(&self) -> bool {
        self.operation_type == OperationType::Query
    }

    pub fn as_nested(&self) -> &[Field<Nested<Input>, Input>] {
        &self.nested
    }

    pub fn into_nested(self) -> Vec<Field<Nested<Input>, Input>> {
        self.nested
    }

    pub fn as_parent(&self) -> &[Field<Flat, Input>] {
        &self.flat
    }

    pub fn find_field(&self, id: FieldId) -> Option<&Field<Flat, Input>> {
        self.flat.iter().find(|field| field.id == id)
    }

    pub fn find_field_path<S: AsRef<str>>(&self, path: &[S]) -> Option<&Field<Flat, Input>> {
        match path.split_first() {
            None => None,
            Some((name, path)) => {
                let field = self.flat.iter().find(|field| field.name == name.as_ref())?;
                if path.is_empty() {
                    Some(field)
                } else {
                    self.find_field_path(path)
                }
            }
        }
    }

    pub fn size(&self) -> usize {
        self.flat.len()
    }
}
