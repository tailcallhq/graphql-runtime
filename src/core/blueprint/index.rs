use indexmap::IndexMap;

use super::InputObjectTypeDefinition;
use crate::core::blueprint::{
    Blueprint, Definition, FieldDefinition, InputFieldDefinition, SchemaDefinition,
};
use crate::core::scalar;

///
/// A read optimized index of all the types in the Blueprint. Provide O(1)
/// access to getting any field information.

#[derive(Debug)]
pub struct Index {
    map: IndexMap<String, (Definition, IndexMap<String, QueryField>)>,
    schema: SchemaDefinition,
}

#[derive(Debug)]
pub enum QueryField {
    Field((FieldDefinition, IndexMap<String, InputFieldDefinition>)),
    InputField(InputFieldDefinition),
}

impl QueryField {
    pub fn get_arg(&self, arg_name: &str) -> Option<&InputFieldDefinition> {
        match self {
            QueryField::Field((_, args)) => args.get(arg_name),
            QueryField::InputField(_) => None,
        }
    }
}

impl Index {
    pub fn type_is_scalar(&self, type_name: &str) -> bool {
        let def = self.map.get(type_name).map(|(def, _)| def);

        matches!(def, Some(Definition::Scalar(_))) || scalar::Scalar::is_predefined(type_name)
    }

    pub fn type_is_enum(&self, type_name: &str) -> bool {
        let def = self.map.get(type_name).map(|(def, _)| def);

        matches!(def, Some(Definition::Enum(_)))
    }

    pub fn validate_enum_value(&self, type_name: &str, value: &str) -> bool {
        let def = self.map.get(type_name).map(|(def, _)| def);

        if let Some(Definition::Enum(enum_)) = def {
            enum_.enum_values.iter().any(|v| v.name == value)
        } else {
            false
        }
    }

    pub fn get_field(&self, type_name: &str, field_name: &str) -> Option<&QueryField> {
        self.map
            .get(type_name)
            .and_then(|(_, fields_map)| fields_map.get(field_name))
    }

    pub fn get_query(&self) -> &String {
        &self.schema.query
    }

    pub fn get_mutation(&self) -> Option<&str> {
        self.schema.mutation.as_deref()
    }

    pub fn is_type_implements(&self, type_name: &str, type_or_interface: &str) -> bool {
        if type_name == type_or_interface {
            return true;
        }

        if let Some((Definition::Object(obj), _)) = self.map.get(type_name) {
            obj.implements.contains(type_or_interface)
        } else {
            false
        }
    }

    pub fn get_input_type_definition(&self, type_name: &str) -> Option<&InputObjectTypeDefinition> {
        match self.map.get(type_name) {
            Some((Definition::InputObject(input), _)) => Some(input),
            _ => None,
        }
    }
}

impl From<&Blueprint> for Index {
    fn from(blueprint: &Blueprint) -> Self {
        let mut map = IndexMap::new();

        for definition in blueprint.definitions.iter() {
            match definition {
                Definition::Object(object_def) => {
                    let type_name = object_def.name.clone();
                    let mut fields_map = IndexMap::new();

                    for field in &object_def.fields {
                        let args_map = IndexMap::from_iter(
                            field
                                .args
                                .iter()
                                .map(|v| (v.name.clone(), v.clone()))
                                .collect::<Vec<_>>(),
                        );
                        fields_map.insert(
                            field.name.clone(),
                            QueryField::Field((field.clone(), args_map)),
                        );
                    }

                    map.insert(
                        type_name,
                        (Definition::Object(object_def.to_owned()), fields_map),
                    );
                }
                Definition::Interface(interface_def) => {
                    let type_name = interface_def.name.clone();
                    let mut fields_map = IndexMap::new();

                    for field in interface_def.fields.clone() {
                        let args_map = IndexMap::from_iter(
                            field
                                .args
                                .iter()
                                .map(|v| (v.name.clone(), v.clone()))
                                .collect::<Vec<_>>(),
                        );
                        fields_map.insert(field.name.clone(), QueryField::Field((field, args_map)));
                    }

                    map.insert(
                        type_name,
                        (Definition::Interface(interface_def.to_owned()), fields_map),
                    );
                }
                Definition::InputObject(input_object_def) => {
                    let type_name = input_object_def.name.clone();
                    let mut fields_map = IndexMap::new();

                    for field in input_object_def.fields.clone() {
                        fields_map.insert(field.name.clone(), QueryField::InputField(field));
                    }

                    map.insert(
                        type_name,
                        (
                            Definition::InputObject(input_object_def.to_owned()),
                            fields_map,
                        ),
                    );
                }
                Definition::Scalar(scalar_def) => {
                    let type_name = scalar_def.name.clone();
                    map.insert(
                        type_name.clone(),
                        (Definition::Scalar(scalar_def.to_owned()), IndexMap::new()),
                    );
                }
                Definition::Enum(enum_def) => {
                    let type_name = enum_def.name.clone();
                    map.insert(
                        type_name.clone(),
                        (Definition::Enum(enum_def.to_owned()), IndexMap::new()),
                    );
                }
                Definition::Union(union_def) => {
                    let type_name = union_def.name.clone();
                    map.insert(
                        type_name.clone(),
                        (Definition::Union(union_def.to_owned()), IndexMap::new()),
                    );
                }
            }
        }

        Self { map, schema: blueprint.schema.to_owned() }
    }
}

#[cfg(test)]
mod test {
    use super::Index;
    use crate::core::blueprint::Blueprint;
    use crate::core::config::ConfigModule;
    use crate::include_config;

    fn setup() -> Index {
        let config = include_config!("./fixture/all-constructs.graphql").unwrap();
        let cfg_module = ConfigModule::from(config);
        let blueprint = Blueprint::try_from(&cfg_module).unwrap();

        Index::from(&blueprint)
    }

    #[test]
    fn test_from_blueprint() {
        let index = setup();
        insta::assert_debug_snapshot!(index);
    }

    #[test]
    fn test_is_scalar() {
        let index = setup();
        assert!(index.type_is_scalar("Int"));
        assert!(index.type_is_scalar("String"));

        assert!(!index.type_is_scalar("Color"));
    }

    #[test]
    fn test_is_enum() {
        let index = setup();
        assert!(index.type_is_enum("Status"));
        assert!(!index.type_is_enum("Int"));
    }

    #[test]
    fn test_validate_enum_value() {
        let index = setup();
        assert!(index.validate_enum_value("Status", "ACTIVE"));
        assert!(!index.validate_enum_value("Status", "YELLOW"));
        assert!(!index.validate_enum_value("Int", "1"));
    }

    #[test]
    fn test_get_field() {
        let index = setup();
        assert!(index.get_field("Query", "user").is_some());
        assert!(index.get_field("Query", "non_existent_field").is_none());
        assert!(index.get_field("Status", "Pending").is_none());
    }

    #[test]
    fn test_get_query() {
        let index = setup();
        assert_eq!(index.get_query(), "Query");
    }

    #[test]
    fn test_get_mutation() {
        let index = setup();
        assert_eq!(index.get_mutation(), Some("Mutation"));
    }

    #[test]
    fn test_get_mutation_none() {
        let mut index = setup();
        index.schema.mutation = None;
        assert_eq!(index.get_mutation(), None);
    }

    #[test]
    fn test_is_type_implements() {
        let index = setup();

        assert!(index.is_type_implements("User", "Node"));
        assert!(index.is_type_implements("Post", "Post"));
        assert!(!index.is_type_implements("Node", "User"));
    }
}
