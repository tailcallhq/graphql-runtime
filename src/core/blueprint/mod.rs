mod auth;
mod blueprint;
mod compress;
mod cors;
mod definitions;
mod dynamic_value;
mod from_config;
mod index;
mod into_schema;
mod links;
mod mustache;
mod operators;
mod schema;
mod server;
pub mod telemetry;
mod timeout;
mod union_resolver;
mod upstream;

pub use auth::*;
pub use blueprint::*;
pub use cors::*;
pub use definitions::*;
pub use dynamic_value::*;
pub use from_config::*;
pub use index::*;
pub use links::*;
pub use operators::*;
pub use schema::*;
pub use server::*;
pub use timeout::GlobalTimeout;
pub use upstream::*;

use crate::core::config::ConfigModule;
use crate::core::try_fold::TryFold;

pub type TryFoldConfig<'a, A> = TryFold<'a, ConfigModule, A, String>;
