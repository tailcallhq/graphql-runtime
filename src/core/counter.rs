use std::cell::Cell;

#[allow(unused)]
#[derive(Default)]
pub struct Counter(Cell<usize>);
impl Counter {
    pub fn new(start: usize) -> Self {
        Self(Cell::new(start))
    }
    pub fn next(&self) -> usize {
        let curr = self.0.get();
        self.0.set(curr + 1);
        curr
    }
}
