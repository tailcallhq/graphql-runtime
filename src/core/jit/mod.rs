mod builder;
mod eval_ir;
mod model;
mod store;
mod synth;
pub use builder::*;
pub use eval_ir::*;
pub use model::*;
pub use store::*;
pub use synth::*;

// NOTE: Only used in tests and benchmarks
pub mod common;
pub mod ir;
