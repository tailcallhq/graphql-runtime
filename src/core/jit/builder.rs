use std::collections::HashMap;
use std::ops::Deref;

use async_graphql::parser::types::{
    Directive, DocumentOperations, ExecutableDocument, FragmentDefinition, OperationType,
    Selection, SelectionSet,
};
use async_graphql::Positioned;
use async_graphql_value::Value;

use super::model::*;
use crate::core::blueprint::{Blueprint, Index, QueryField};
use crate::core::counter::{Count, Counter};
use crate::core::merge_right::MergeRight;

#[derive(PartialEq)]
enum Condition {
    True,
    False,
    Variable(Variable),
}

struct Conditions {
    skip: Option<Condition>,
    include: Option<Condition>,
}

impl Conditions {
    /// Checks if the field should be skipped always
    fn is_const_skip(&self) -> bool {
        matches!(self.skip, Some(Condition::True)) ^ matches!(self.include, Some(Condition::True))
    }

    fn into_variable_tuple(self) -> (Option<Variable>, Option<Variable>) {
        let comp = |condition| match condition? {
            Condition::Variable(var) => Some(var),
            _ => None,
        };

        let include = comp(self.include);
        let skip = comp(self.skip);

        (include, skip)
    }
}

pub struct Builder {
    pub index: Index,
    pub arg_id: Counter<usize>,
    pub field_id: Counter<usize>,
    pub document: ExecutableDocument,
}

impl Builder {
    pub fn new(blueprint: &Blueprint, document: ExecutableDocument) -> Self {
        let index = blueprint.index();
        Self {
            document,
            index,
            arg_id: Counter::default(),
            field_id: Counter::default(),
        }
    }

    #[inline(always)]
    fn include(
        &self,
        directives: &[Positioned<async_graphql::parser::types::Directive>],
    ) -> Conditions {
        fn get_condition(dir: &Directive) -> Option<Condition> {
            let arg = dir.get_argument("if").map(|pos| &pos.node);
            let is_include = dir.name.node.as_str() == "include";
            match arg {
                None => None,
                Some(value) => match value {
                    Value::Boolean(bool) => {
                        let condition = if is_include ^ bool {
                            Condition::True
                        } else {
                            Condition::False
                        };
                        Some(condition)
                    }
                    Value::Variable(var) => {
                        Some(Condition::Variable(Variable::new(var.deref().to_owned())))
                    }
                    _ => None,
                },
            }
        }
        Conditions {
            skip: directives
                .iter()
                .find(|d| d.node.name.node.as_str() == "skip")
                .map(|d| &d.node)
                .and_then(get_condition),
            include: directives
                .iter()
                .find(|d| d.node.name.node.as_str() == "include")
                .map(|d| &d.node)
                .and_then(get_condition),
        }
    }

    #[allow(clippy::too_many_arguments)]
    #[inline(always)]
    fn iter(
        &self,
        selection: &SelectionSet,
        type_of: &str,
        exts: Option<Flat>,
        fragments: &HashMap<&str, &FragmentDefinition>,
    ) -> Vec<Field<Flat>> {
        let mut fields = vec![];
        for selection in &selection.items {
            match &selection.node {
                Selection::Field(Positioned { node: gql_field, .. }) => {
                    let conditions = self.include(&gql_field.directives);

                    // if include is always false xor skip is always true,
                    // then we can skip the field from the plan
                    if conditions.is_const_skip() {
                        continue;
                    }

                    let (include, skip) = conditions.into_variable_tuple();

                    let field_name = gql_field.name.node.as_str();
                    let field_args = gql_field
                        .arguments
                        .iter()
                        .map(|(k, v)| (k.node.as_str().to_string(), v.node.to_owned()))
                        .collect::<HashMap<_, _>>();

                    if let Some(field_def) = self.index.get_field(type_of, field_name) {
                        let mut args = vec![];
                        for (arg_name, value) in field_args {
                            if let Some(arg) = field_def.get_arg(&arg_name) {
                                let type_of = arg.of_type.clone();
                                let id = ArgId::new(self.arg_id.next());
                                let name = arg_name.clone();
                                let default_value = arg
                                    .default_value
                                    .as_ref()
                                    .and_then(|v| v.to_owned().try_into().ok());
                                args.push(Arg {
                                    id,
                                    name,
                                    type_of,
                                    value: Some(value),
                                    default_value,
                                });
                            }
                        }

                        let type_of = match field_def {
                            QueryField::Field((field_def, _)) => field_def.of_type.clone(),
                            QueryField::InputField(field_def) => field_def.of_type.clone(),
                        };

                        let id = FieldId::new(self.field_id.next());
                        let child_fields = self.iter(
                            &gql_field.selection_set.node,
                            type_of.name(),
                            Some(Flat::new(id.clone())),
                            fragments,
                        );
                        let name = field_name.to_owned();
                        let ir = match field_def {
                            QueryField::Field((field_def, _)) => field_def.resolver.clone(),
                            _ => None,
                        };
                        fields.push(Field {
                            id,
                            name,
                            ir,
                            type_of,
                            skip,
                            include,
                            args,
                            extensions: exts.clone(),
                        });
                        fields = fields.merge_right(child_fields);
                    }
                }
                Selection::FragmentSpread(Positioned { node: fragment_spread, .. }) => {
                    if let Some(fragment) =
                        fragments.get(fragment_spread.fragment_name.node.as_str())
                    {
                        fields.extend(self.iter(
                            &fragment.selection_set.node,
                            fragment.type_condition.node.on.node.as_str(),
                            exts.clone(),
                            fragments,
                        ));
                    }
                }
                _ => {}
            }
        }

        fields
    }

    #[inline(always)]
    fn get_type(&self, ty: OperationType) -> Option<&str> {
        match ty {
            OperationType::Query => Some(self.index.get_query()),
            OperationType::Mutation => self.index.get_mutation(),
            OperationType::Subscription => None,
        }
    }

    #[inline(always)]
    pub fn build(&self) -> Result<ExecutionPlan, String> {
        let mut fields = Vec::new();
        let mut fragments: HashMap<&str, &FragmentDefinition> = HashMap::new();

        for (name, fragment) in self.document.fragments.iter() {
            fragments.insert(name.as_str(), &fragment.node);
        }

        match &self.document.operations {
            DocumentOperations::Single(single) => {
                let name = self.get_type(single.node.ty).ok_or(format!(
                    "Root Operation type not defined for {}",
                    single.node.ty
                ))?;
                fields.extend(self.iter(&single.node.selection_set.node, name, None, &fragments));
            }
            DocumentOperations::Multiple(multiple) => {
                for single in multiple.values() {
                    let name = self.get_type(single.node.ty).ok_or(format!(
                        "Root Operation type not defined for {}",
                        single.node.ty
                    ))?;
                    fields.extend(self.iter(
                        &single.node.selection_set.node,
                        name,
                        None,
                        &fragments,
                    ));
                }
            }
        }

        Ok(ExecutionPlan::new(fields))
    }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;

    use super::*;
    use crate::core::blueprint::Blueprint;
    use crate::core::config::Config;
    use crate::core::jit::builder::Builder;
    use crate::core::valid::Validator;

    const CONFIG: &str = include_str!("./fixtures/jsonplaceholder-mutation.graphql");

    fn plan(query: impl AsRef<str>) -> ExecutionPlan {
        let config = Config::from_sdl(CONFIG).to_result().unwrap();
        let blueprint = Blueprint::try_from(&config.into()).unwrap();
        let document = async_graphql::parser::parse_query(query).unwrap();
        Builder::new(&blueprint, document).build().unwrap()
    }

    #[tokio::test]
    async fn test_from_document() {
        let plan = plan(
            r#"
            query {
                posts { user { id name } }
            }
        "#,
        );
        insta::assert_debug_snapshot!(plan.into_nested());
    }

    #[tokio::test]
    async fn test_size() {
        let plan = plan(
            r#"
            query {
                posts { user { id name } }
            }
        "#,
        );

        assert_eq!(plan.size(), 4)
    }

    #[test]
    fn test_simple_query() {
        let plan = plan(
            r#"
            query {
                posts { user { id } }
            }
        "#,
        );
        insta::assert_debug_snapshot!(plan.into_nested());
    }

    #[test]
    fn test_simple_mutation() {
        let plan = plan(
            r#"
            mutation {
              createUser(user: {
                id: 101,
                name: "Tailcall",
                email: "tailcall@tailcall.run",
                phone: "2345234234",
                username: "tailcall",
                website: "tailcall.run"
              }) {
                id
                name
                email
                phone
                website
                username
              }
            }
        "#,
        );
        insta::assert_debug_snapshot!(plan.into_nested());
    }

    #[test]
    fn test_fragments() {
        let plan = plan(
            r#"
            fragment UserPII on User {
              name
              email
              phone
            }

            query {
              user(id:1) {
                ...UserPII
              }
            }
        "#,
        );
        insta::assert_debug_snapshot!(plan.into_nested());
    }

    #[test]
    fn test_multiple_operations() {
        let plan = plan(
            r#"
            query {
              user(id:1) {
                id
                username
              }
              posts {
                id
                title
              }
            }
        "#,
        );
        insta::assert_debug_snapshot!(plan.into_nested());
    }

    #[test]
    fn test_variables() {
        let plan = plan(
            r#"
            query user($id: Int!) {
              user(id: $id) {
                id
                name
              }
            }
        "#,
        );
        insta::assert_debug_snapshot!(plan.into_nested());
    }

    #[test]
    fn test_unions() {
        let plan = plan(
            r#"
            query {
              getUserIdOrEmail(id:1) {
                ...on UserId {
                  id
                }
                ...on UserEmail {
                  email
                }
              }
            }
        "#,
        );
        insta::assert_debug_snapshot!(plan.into_nested());
    }

    #[test]
    fn test_default_value() {
        let plan = plan(
            r#"
            mutation {
              createPost(post:{
                userId:123,
                title:"tailcall",
                body:"tailcall test"
              }) {
                id
              }
            }
        "#,
        );
        insta::assert_debug_snapshot!(plan.into_nested());
    }
}
