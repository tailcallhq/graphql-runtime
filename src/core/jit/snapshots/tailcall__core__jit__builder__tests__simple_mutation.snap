---
source: src/core/jit/builder.rs
expression: plan.selection
snapshot_kind: text
---
[
    Field {
        id: 0,
        name: "createUser",
        output_name: "createUser",
        ir: "Some(..)",
        type_of: User,
        type_condition: Some(
            "Mutation",
        ),
        args: [
            Arg {
                id: 0,
                name: "user",
                type_of: InputUser!,
                value: Some(
                    Object(
                        {
                            Name(
                                "id",
                            ): Number(
                                Number(101),
                            ),
                            Name(
                                "name",
                            ): String(
                                "Tailcall",
                            ),
                            Name(
                                "email",
                            ): String(
                                "tailcall@tailcall.run",
                            ),
                            Name(
                                "phone",
                            ): String(
                                "2345234234",
                            ),
                            Name(
                                "username",
                            ): String(
                                "tailcall",
                            ),
                            Name(
                                "website",
                            ): String(
                                "tailcall.run",
                            ),
                        },
                    ),
                ),
                default_value: None,
            },
        ],
        selection: [
            Field {
                id: 1,
                name: "id",
                output_name: "id",
                type_of: ID!,
                type_condition: Some(
                    "User",
                ),
                directives: [],
            },
            Field {
                id: 2,
                name: "name",
                output_name: "name",
                type_of: String!,
                type_condition: Some(
                    "User",
                ),
                directives: [],
            },
            Field {
                id: 3,
                name: "email",
                output_name: "email",
                type_of: String!,
                type_condition: Some(
                    "User",
                ),
                directives: [],
            },
            Field {
                id: 4,
                name: "phone",
                output_name: "phone",
                type_of: String,
                type_condition: Some(
                    "User",
                ),
                directives: [],
            },
            Field {
                id: 5,
                name: "website",
                output_name: "website",
                type_of: String,
                type_condition: Some(
                    "User",
                ),
                directives: [],
            },
            Field {
                id: 6,
                name: "username",
                output_name: "username",
                type_of: String!,
                type_condition: Some(
                    "User",
                ),
                directives: [],
            },
        ],
        directives: [],
    },
]
